### DISCUSSION TITLE
AMP measurement support

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#220] [Anything Else]
"My website supports AMP. Given that AMP is growing, measuring this would be key for me. The AMP standard is an open one. There were more than 55 analytics providers supporting AMP even in 2018, like Amplitude or Adobe Analytics or New Relic. [Here's a link](https://instapage.com/blog/amp-analytics) with a bunch of the AMP API spec for many providers."

feedback via email

**Comment by gergelyorosz** [comment]
There are now 75 analytics vendors who are compatible with AMP (full list is [here](https://github.com/rudderlabs/amphtml/blob/1c7cf35b91eae727897d9c7d3448fa9a6c7f5fde/examples/analytics-vendors.amp.html)).

Here is the documentation on how to be added as an AMP vendor: https://amp.dev/documentation/components/amp-analytics/ . Here's an [example PR for RudderStack ](https://github.com/ampproject/amphtml/pull/28597/files)on being added as a vendor or one [for BlueConic](https://github.com/ampproject/amphtml/pull/28102/files).

**Comment by tckb** [comment]
@gergelyorosz  I plan to add this in my next PR.  Keep a watch

**Comment by davidsteeb** [comment]
Any news on this? If you need support in testing let me know, eager to try this.

**Comment by HopsWrites** [comment]
Hey Plausible,

After inspecting my main URL on Search Console, it said the “AMP Page is invalid” due to custom Javascript not being allowed. The only custom script I have is Plausible's.

Google is moving to indexing only mobile-responsive pages.

Any updates on AMP and Plausible?

Thanks, and LOVE Plausible so far!  (-:

**Comment by metmarkosaric** [comment]
Thanks, glad to hear that! We're about to release metadata for goals and after that [our roadmap](https://github.com/plausible/analytics/projects/1) is focused on other popular requests (browser/os versions, API, export/import). After that we'll consider what to do next and see if we can put AMP on the roadmap at that point. Obviously if there's a PR before, we'll happily merge it. You don't have a native mobile responsive page but only an AMP?

**Comment by HopsWrites** [comment]
Thanks for getting back!

Well, I revised my blog using Publii, which offers code injection for both
the main site (which is mobile responsive), and AMP. I had placed Plausible
code in both Headers, but Google Search Console said the custom Javascript
invalidated my AMP page. No idea why. I removed the code from AMP for now.
My blog is at blog.hopswrites.com

Thanks for your time.

-Jordan

**Comment by metmarkosaric** [comment]
Makes sense. Can see that your blog is mobile responsive on its own too. For AMP, we'll need to be added as a vendor before Plausible can be used on AMP pages.

**Comment by HopsWrites** [comment]
Thanks for checking and replying!

**Comment by kingpalethe** [comment]
This is something we need also. AMP is totally anti-competitive, and is one of the reasons the U.S. States are suing Google, but until Google is brought into line, we need to play by their rules, which basically require AMP.

**Comment by pxDot** [comment]
I agree. AMP is sadly a must have. Is there any workaround on how we can push some basic metrics into plausible from AMP pages in the meantime?

This is currently a showstopper, especially for larger content websites where almost 90% of traffic is hitting the AMP pages.

**Reply by metmarkosaric** [reply]
thanks @squareloop1! there's no current plan to work on the AMP support. we're a small team so need to prioritize. we're focused on the API at this moment as it's the most popular requested feature. after that it's probably going to be import/export as it's the second most popular request. not sure if there might be a way to send AMP views through our API? i know it will be possible to do so for the mobile app views so perhaps something like that works for AMP too? something we can look into

**Reply by pxDot** [reply]
The API might actually work to manually push view / click events from AMP pages to plausible. I'm looking forward to test this in AMP. Thank you.

**Reply by metmarkosaric** [reply]
that would be amazing! do keep an eye on our API beta release details and hopefully you can help us confirm this use case.

**Reply by pxDot** [reply]
I will definetly look into this as soon as the API beta is out and will let you know if I can fiddle something together in AMP that works with plausible. Is there any ETA on the beta?

**Reply by metmarkosaric** [reply]
cool! we don't have any ETA but as mentioned we're working on it right now and it is priority number 1. stay tuned!

**Reply by ukutaht** [reply]
I'm aiming for next week @squareloop1

**Reply by pxDot** [reply]
Is there any news on this? I've seen there is now an API available but only to read and not to push data into it.

**Reply by metmarkosaric** [reply]
the beta version of our Stats API is now out and we're collecting feedback from users before releasing version 1.0. in the meanwhile we're focusing on some other frequently requested features such as entry/exit pages and user accounts. when the Stats API hits 1.0 we will work on 2.0 version which will also allow you to send pageviews/goals and not only read. i cannot promise any timeline on this as we're taking it one step at a time.

**Comment by Bammede** [comment]
Is it possible to generate Plausible's tracking code for AMP pages already?

**Comment by pxDot** [comment]
Now that the API is in the 1.3.0 release, is there any chance the next step is an API endpoint to push custom data from AMP pages into plausible or any other method to track AMP pages with plausible?

I really want to migrate all sites to plausible but its just not possible while it still does not support at least basic tracking on AMP pages too.

**Comment by metmarkosaric** [comment]
From next month, there will no longer be any preferential treatment for AMP pages in Google's search results! https://plausible.io/blog/google-amp

**Reply by pxDot** [reply]
Yes, this is a big move to finally get rid of AMP again. But I would say we still need at least basic AMP tracking in plausible as it won't get erased instantly from the internet in the next month as many publishers already have a working AMP setup.

As many publishers already have AMP implemented anyways and google will still prefer mobile pages by its performance you'd need to make your non-AMP page is at least as fast as the AMP version to profit from the new ranking algorithm, which is not as simple as it might first look like. AMP is a pain in the ass, I agree, but it is also very performant, making it hard to outperform on a non-AMP page.

Also plausible already has the needed architecture available and implementing basic AMP tracking should not be a big issue. As you can read in the docs (https://amp.dev/documentation/components/amp-analytics/) we would just need to define data getting send to an API endpoint in plausible.

I would devote some time to it and create a PR but sadly I am unfamiliar with Elixir.

**Comment by pxDot** [comment]
After finally putting some time into this I figured out it is totally doable to track AMP with plausible already. We don't need to implement any new feature for this. Maybe we could just mention it in the docs on how to do this.

To track AMP request and outbound links with plausible you need to declare an amp-analytics object on your AMP pages like this:

```html
<amp-analytics>
    <script type="application/json">
        {
            "vars": {
                "dataDomain": "example.com"
            },
            "requests": {
                "event": "https://plausible.io/api/event"
            },
            "extraUrlParams": {
                "u": "${sourceUrl}",
                "r": "${documentReferrer}",
                "w": "${viewportWidth}",
                "d": "${dataDomain}"
            },
            "triggers": {
                "trackPageview": {
                    "on": "visible",
                    "request": "event",
                    "extraUrlParams": {
                        "n": "pageview"
                    }
                },
                "trackOutboundLinks": {
                    "on": "click",
                    "request": "event",
                    "selector": "a[href*='//']:not([href*='${sourceHost}'])",
                    "extraUrlParams": {
                        "n": "Outbound Link: Click",
                        "p": "{\"url\":\"${clickUrl}\"}"
                    }
                }
            },
            "transport": {
                "beacon": true,
                "xhrpost": true,
                "image": false,
                "useBody": true
            }
        }
    </script>
</amp-analytics>
```

The `${clickUrl}` variable needs to be set on outbound links as [data attribute](https://github.com/ampproject/amphtml/blob/main/extensions/amp-analytics/analytics-vars.md#variables-as-data-attribute). You can also add other events to define custom goals. More details on this can be found here: https://amp.dev/documentation/components/amp-analytics/#send-data-in-house

**Reply by metmarkosaric** [reply]
that's great, thanks for sharing @squareloop1! i've added a note and a link to your instructions in our integrations list https://plausible.io/docs/integration-guides#google-amp-pages

**Reply by iamkingsleyf** [reply]
@squareloop1 this is a very long code, any idea how to use it on the official AMP plugin

**Reply by iamkingsleyf** [reply]
At first it did not work because of the script tags... now it tracks the main URL and not the AMP.
official support will be nice @ukutaht @metmarkosaric even if AMP is dead to you guys (i have seen your blog post)

**Reply by pxDot** [reply]
@iamkingsleyf I don't know what you mean with "official AMP plugin". You should only add this code to your AMP pages. It should not track your normal pages then because of the special `<amp-analytics>` tag used only on AMP pages.

Please have a look at the official docs: https://amp.dev/documentation/components/amp-analytics/#send-data-in-house on how to implement this.

There is really no need for "official support" because there is nothing to implement here. It is working fine as it is. You can say plausible "officially supports AMP" already 😅

**Reply by iamkingsleyf** [reply]
It didn't work with that amp-analytics script

**Reply by pxDot** [reply]
The amp-analytics script tag works fine if you put it in the HTML of the AMP page itself.

I am assuming you are using some kind of Wordpress plugin or something as you talked about an "official AMP plugin". In that case there might be some differences in implementation. I don't think this is in the scope of plausible and you should probably ask for support from the plugin devs on how to implement custom AMP analytics code.

**Reply by iamkingsleyf** [reply]
We use the official AMP plugin for Wordpress and here is how the code was able to work https://imgur.com/gUp316w it did not accept <amp-analytics>
    <script type="application/json"> when i tried adding all.

what platform are you using

**Reply by iamkingsleyf** [reply]
From the source code those script part is being added, but the problem is plausible does not track the amp URL just the normal url

**Reply by pawelslabiak** [reply]
Thank you @squareloop1 for sharing this here and @metmarkosaric for including this in Plausible's documentation. I run a WP website and use Plausible's WP plugin, but needed to extend it to track posts published via Google's Web Stories plugin, which uses AMP. 

@squareloop1's code, combined with code examples by @swissspidy (see https://wordpress.org/support/topic/analytics-not-ga/) gave me the solution. I run this via the Code Snippets plugin, although it could also be implemented as a site-specific plugin.

In case it's useful for anyone else, here's the complete snippet: 

```
<?php
add_action(
	'web_stories_print_analytics',
	static function () {
		?>
		<amp-analytics>
            <script type="application/json"> {
                "vars": {
                    "dataDomain": "yourdomain.com"
                },
                "requests": {
                    "event": "https://plausible.io/api/event"
                },
                "extraUrlParams": {
                    "u": "${sourceUrl}",
                    "r": "${documentReferrer}",
                    "w": "${viewportWidth}",
                    "d": "${dataDomain}"
                },
                "triggers": {
                    "trackPageview": {
                        "on": "visible",
                        "request": "event",
                        "extraUrlParams": {
                            "n": "pageview"
                        }
                    },
                    "trackOutboundLinks": {
                        "on": "click",
                        "request": "event",
                        "selector": "a[href*='//']:not([href*='${sourceHost}'])",
                        "extraUrlParams": {
                            "n": "Outbound Link: Click",
                            "p": "{\"url\":\"${clickUrl}\"}"
                        }
                    }
                },
                "transport": {
                    "beacon": true,
                    "xhrpost": true,
                    "image": false,
                    "useBody": true
                }
            }
        </script>
    </amp-analytics>
    <?php
    }
);
```

Remember to replace "yourdomain.com" with your Plausible data domain. 

The one wrinkle here is outbound-link tracking. @squareloop1 mentions that 

> The ${clickUrl} variable needs to be set on outbound links as [data attribute](https://github.com/ampproject/amphtml/blob/main/extensions/amp-analytics/analytics-vars.md#variables-as-data-attribute).

I'm not sure what this means and haven't done any additional configuration besides running the code shown above, hence the outbound links on web stories are not tracked on the site in question. Would love to understand better how to make this work.

**Reply by pxDot** [reply]
>> The ${clickUrl} variable needs to be set on outbound links as data attribute.

> I'm not sure what this means and haven't done any additional configuration besides running the code shown above, hence the outbound links on web stories are not tracked on the site in question. Would love to understand better how to make this work.

Its been a while since I've implemented this, but I think the way is to set a `data-vars-click-url` attribute on outbound link tags. So e.g. like this:
`<a href="https://google.com" data-vars-click-url="https://google.com">`

and the variable `${clickUrl}` will get populated with the corresponding URL and it will track the outbound link.

**Reply by pawelslabiak** [reply]
Awesome—thank you, @squareloop1, for explaining this!

**Comment by kpnigalye** [comment]
@squareloop1  The script worked for me. I have noticed one thing. I opened my site on my mobile browser and its still showed up under desktop device. Its using `${viewportWidth}` for calculating width. Not sure why its showing the screen size as desktop. But it identifies the device properly.

**Reply by dynamicplus** [reply]
That's why the "w" (or "screen_width") parameter is not casted to an integer if passed as numeric string.
